examples/auction/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+                Ok(return_action)
+            }
+            None =>
+                bail!(FinalizeError::BidMapError)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use test_infrastructure::*;
+    use std::sync::atomic::{AtomicU8, Ordering};
+
+    // A counter for generating new account addresses
+    static ADDRESS_COUNTER: AtomicU8 = AtomicU8::new(0);
+    const AUCTION_END: u64 = 1;
+    const ITEM: &str = "Starry night by Van Gogh";
+
+    fn dummy_fresh_state() -> State {
+        dummy_state(Amount::zero(), BTreeMap::new())
+    }
+
+    fn dummy_state(highest: Amount, bids: BTreeMap<AccountAddress, Amount>) -> State {
+        State {
+            auction_state: AuctionState::Active,
+            highest_bid: highest,
+            item: ITEM.as_bytes().to_vec(),
+            expiry: Timestamp::from_timestamp_millis(AUCTION_END),
+            bids: bids,
+        }
+    }
+
+    fn expect_error<E, T>(expr: Result<T, E>, err: E, msg: &str)
+        where E: Eq + Debug, T: Debug {
+        let actual = expr.expect_err(msg);
+        assert_eq!(actual, err);
+    }
+
+    fn item_expiry_parameter() -> InitParameter {
+        InitParameter {
+            item: ITEM.as_bytes().to_vec(),
+            expiry: Timestamp::from_timestamp_millis(AUCTION_END),
+        }
+    }
+
+    fn create_parameter_bytes(parameter: &InitParameter) -> Vec<u8> {
+        to_bytes(parameter)
+    }
+
+    fn parametrized_init_ctx<'a>(parameter_bytes: &'a Vec<u8>) -> InitContextTest<'a> {
+        let mut ctx = InitContextTest::empty();
+        ctx.set_parameter(parameter_bytes);
+        ctx
+    }
+
+    fn new_account() -> AccountAddress {
+        let account = AccountAddress([ADDRESS_COUNTER.load(Ordering::SeqCst); 32]);
+        ADDRESS_COUNTER.fetch_add(1, Ordering::SeqCst);
+        account
+    }
+
+    fn new_account_ctx<'a>() -> (AccountAddress, ReceiveContextTest<'a>) {
+        let account = new_account();
+        let ctx = new_ctx(account, account, AUCTION_END);
+        (account, ctx)
+    }
+
+    fn new_ctx<'a>(owner: AccountAddress, sender: AccountAddress, slot_time: u64) -> ReceiveContextTest<'a> {
+        let mut ctx = ReceiveContextTest::empty();
+        ctx.set_sender(Address::Account(sender));
+        ctx.set_owner(owner);
+        ctx.set_metadata_slot_time(Timestamp::from_timestamp_millis(slot_time));
+        ctx
+    }
+
+    #[test]
+    fn test_init() {
+        let parameter_bytes = create_parameter_bytes(&item_expiry_parameter());
+        let ctx = parametrized_init_ctx(&parameter_bytes);
+
+        let state_result = auction_init(&ctx);
+        let state = state_result.expect("Contract initialization results in error");
+        assert_eq!(state,
+                   dummy_fresh_state(),
+                   "Auction state should be new after initialization"
+        );
+    }
+
+    #[test]
+    fn test_auction_bid_and_finalize() {
```

Comment:
I would prefer if this was split into multiple unit tests, since it testing more than one thing, meaning if any of them failed it would be hard to know what exactly failed.

Reply:
I disagree with this. I like this test. It is testing a lot of failed and successful scenarios in sequence, also testing that things can be done after failed attempts.

Moreover there are concrete error messages attached to each test, which will make it easy to identify them when they fail.